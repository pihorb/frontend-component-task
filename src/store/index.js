export {context} from './context';
export {reducer} from './reducer';
export {types} from './types';
