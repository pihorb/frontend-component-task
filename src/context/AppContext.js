import React, {createContext, useContext, useReducer} from 'react';
import reducer from '../reducers/reducer';
import {ADD_FOLDERS, ADD_GEMS, REMOVE_FOLDERS, REMOVE_GEMS} from '../reducers/types';

const AppContext = createContext();

export const useAppContext = () => {
  return useContext(AppContext);
};

export const setCheckBoxes = (...opts) => {
  const min = ['view', 'share'];

  if (opts) min.push(...opts);

  return min;
};

export const AppProvider = ({children}) => {
  const [state, dispatch] = useReducer(reducer, {
    role: 'user',
    lastChoice: null,
    checkBoxes: {
      gems: setCheckBoxes('create', 'update'),
      folders: setCheckBoxes('create', 'update')
    }
  });

  const setActiveRole = (role) => {
    dispatch({type: role.name.toUpperCase()});
  };

  const updateCheckBoxes = (name, path, isChecked) => {
    switch (true) {
      case isChecked && path === 'gems':
        return dispatch({type: REMOVE_GEMS, payload: name});
      case isChecked && path === 'folders':
        return dispatch({type: REMOVE_FOLDERS, payload: name});
      case path === 'folders':
        return dispatch({type: ADD_FOLDERS, payload: name});
      case path === 'gems':
        return dispatch({type: ADD_GEMS, payload: name});
      default:
        return null;
    }
  };

  return (
    <AppContext.Provider
      value={{
        toggle: setActiveRole,
        updateCheckBoxes,
        role: state.role,
        checkBoxes: state.checkBoxes,
        lastChoice: state.lastChoice
      }}
    >
      {children}
    </AppContext.Provider>
  );
};
