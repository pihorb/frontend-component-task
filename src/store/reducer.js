import {
  ADD_FOLDERS,
  ADD_GEMS,
  ADMIN,
  CUSTOM,
  MEMBER,
  REMOVE_FOLDERS,
  REMOVE_GEMS,
  USER
} from './types';
import {defaultSelected} from './context';

const reducer = (state, action) => {
  switch (action.type) {
    case USER:
      return {
        ...state,
        role: 'user',
        slots: {
          gems: defaultSelected('create', 'update'),
          folders: defaultSelected('create', 'update')
        }
      };
    case MEMBER:
      return {
        ...state,
        role: 'member',
        slots: {
          gems: defaultSelected(),
          folders: defaultSelected()
        }
      };
    case ADMIN:
      return {
        ...state,
        role: 'admin',
        slots: {
          gems: defaultSelected('create', 'update', 'move', 'delete'),
          folders: defaultSelected('create', 'update', 'move', 'delete')
        }
      };
    case CUSTOM:
      return {
        ...state,
        role: 'custom',
        slots: {
          gems: [],
          folders: []
        }
      };
    case ADD_GEMS:
      return {
        ...state,
        lastChoice: action.payload,
        slots: {
          gems: [...state.slots.gems, action.payload],
          folders: [...state.slots.folders]
        }
      };
    case ADD_FOLDERS:
      return {
        ...state,
        lastChoice: action.payload,
        slots: {
          gems: [...state.slots.gems],
          folders: [...state.slots.folders, action.payload]
        }
      };
    case REMOVE_GEMS:
      return {
        ...state,
        slots: {
          gems: state.slots.gems.filter((item) => item !== action.payload),
          folders: [...state.slots.folders]
        }
      };
    case REMOVE_FOLDERS:
      return {
        ...state,
        slots: {
          gems: [...state.slots.gems],
          folders: state.slots.folders.filter((item) => item !== action.payload)
        }
      };
    default:
      return state;
  }
};

export default reducer;
