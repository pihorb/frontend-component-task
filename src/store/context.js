import React, {createContext, useContext, useReducer} from 'react';
import reducer from './reducer';
import {ADD_FOLDERS, ADD_GEMS, REMOVE_FOLDERS, REMOVE_GEMS} from './types';

const AppContext = createContext();

export const useAppContext = () => {
  return useContext(AppContext);
};

export const defaultSelected = (...opts) => {
  const min = ['view', 'share'];

  if (opts) min.push(...opts);

  return min;
};

export const AppProvider = ({children}) => {
  const [state, dispatch] = useReducer(reducer, {
    role: 'user',
    lastChoice: null,
    slots: {
      gems: defaultSelected('create', 'update'),
      folders: defaultSelected('create', 'update')
    }
  });

  const toggleRole = (role) => {
    dispatch({type: role.name.toUpperCase()});
  };

  const toggleSelect = (name, path, isChecked) => {
    switch (true) {
      case isChecked && path === 'gems':
        return dispatch({type: REMOVE_GEMS, payload: name});
      case isChecked && path === 'folders':
        return dispatch({type: REMOVE_FOLDERS, payload: name});
      case path === 'folders':
        return dispatch({type: ADD_FOLDERS, payload: name});
      case path === 'gems':
        return dispatch({type: ADD_GEMS, payload: name});
      default:
        return null;
    }
  };

  return (
    <AppContext.Provider
      value={{
        toggleRole,
        toggleSelect,
        role: state.role,
        slots: state.slots,
        lastChoice: state.lastChoice
      }}
    >
      {children}
    </AppContext.Provider>
  );
};
