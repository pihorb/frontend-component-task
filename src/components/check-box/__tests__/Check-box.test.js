import renderer from 'react-test-renderer';
import '@testing-library/jest-dom';
import React from 'react';
import {cleanup, fireEvent, render} from '@testing-library/react';
import CheckBox from '../Check-box';
import {AppProvider} from '../../../store/context';

const renderCheckBox = (cb, opts = {name: 'delete', path: 'folders'}) => {
  return cb(
    <AppProvider>
      <CheckBox {...opts} />
    </AppProvider>
  );
};

beforeEach(cleanup);

describe('checkBox component test', () => {
  it('renders correctly', () => {
    const {queryByTestId, getByAltText} = renderCheckBox(render);

    expect(queryByTestId('check-box')).toBeTruthy();
    expect(getByAltText('empty')).toBeTruthy();
    expect(queryByTestId('check-box').children[1]).toBeTruthy();
  });

  it('displays checkbox title', () => {
    const {getByAltText} = renderCheckBox(render);
    const icon = getByAltText('empty');

    fireEvent.click(icon);

    expect(icon.alt).toMatch('checked');
  });

  it('update state on click', () => {
    const {queryByTestId} = renderCheckBox(render, {name: 'view', path: 'gems'});

    expect(queryByTestId('check-box').children[1]).toHaveTextContent('view');
  });

  it('check if snapshot is matching', () => {
    const tree = renderCheckBox(renderer.create).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
