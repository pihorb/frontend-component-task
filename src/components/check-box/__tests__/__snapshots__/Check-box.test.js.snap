// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckBox component test check if snapshot is matching  1`] = `
<div
  className="check-box"
  data-testid="check-box"
  onClick={[Function]}
  style={
    Object {
      "opacity": 0.5,
      "pointerEvents": "none",
    }
  }
>
  <img
    alt="empty"
    className="check-box__img"
    src={Object {}}
  />
  <span
    className="check-box__name"
  >
    delete
  </span>
</div>
`;
